### ISSUE TITLE
Supporting services changes are not in the tested in the pipeline

### ISSUE THREAD

**Issue by joshuawilson** [openshiftio/openshift.io#435]
If a change is done to any of the supporting service they can break the application because they are not being varified by the EE tests. We can run the tests and find that it is broken but that is only after they are deployed.

How do we apply the EE tests to these service before they are deployed?
- [ ] tenant
- [ ] Jenkins
- [ ] Che
- [ ] WIT
- [ ] Auth

**Comment by aslakknutsen** [comment]
Tenant Service will install a new tenant setup for n test users, 1 'fresh', 1 being updated, then kick off the EE tests before it allows the new v to be available for update/install.

As for Auth and WIT and other dsaas services, we should setup a 'shadow' deploy. Essentially have smart routers that can route on content of the Token or similar. That way we can deploy and place a new WIT v live with all the other Live services, but test accounts are routed to this specific shadow route in prod instead of to the live prod instances. The shadow deployment is then promoted to the live route when seen ok. 

Additionally, we need a lot of smart monitoring to look out for changes in trends, e.g. increased number of >299 http response codes, response times etc.

The shadow route style of deployment could even be used for dev in prod. We could technically say that a Branch in a upstream repo becomes a shadow route, and if multiple services use the same branch name they end in the same shadow env. A user could then opt in to see a shadow v of the system, e.g. pm approval or similar on real data unseen by the normal user.

**Comment by alexeykazakov** [comment]
Sounds interesting @aslakknutsen. But we should be careful with these shadow deployments since they will be manipulating the same production data as live deployments (that's the point though :) vs. local development or staging env where screwing the data is not so critical.

**Comment by joshuawilson** [comment]
Another question, what security risks are we opening up with these shadow routes and deployments.

**Comment by joshuawilson** [comment]
I do think it is an interesting way to solve the problem.

**Comment by aslakknutsen** [comment]
> Another question, what security risks are we opening up with these shadow routes and deployments.

The security needs to be around the whole thing anyway. It needs to be a level above.

**Comment by aslakknutsen** [comment]
@alexeykazakov Of course, data migration is a risk, but no more risk than when you do an upgrade. You technically do the same(having multiple instances running on different versions). As far as interacting with other services, that's not the shadow services problem. The service you interact with needs to be tight anyway since others are already interacting with it. 

If e.g. you're afraid that WIT will do something strange when you call it from another 'unstable' service we have a much larger issue as WIT is an open public service called by anyone online essentially. 'The others' will manage to call WIT in strange ways we couldn't even dream up, by accident.

**Comment by alexeykazakov** [comment]
Sure. We should just treat the shadow services are real services working in the prod env. It's just received limited traffic from "test" users. In other words we can do a shadow deployment when we think it's production ready :)

**Comment by ldimaggi** [comment]
+1 to the idea of shadow deployments!

**Comment by bmicklea** [comment]
Who would work on this (I assume it's not yet in progress)?

**Comment by gazarenkov** [comment]
For my understanding:
Testing on staging environment before prod is not an option?
